lib/writable.js

```diff
@@ -1,69 +1,233 @@
+    chunk.copy(this.buffer, this.offset);
+    this.writableLength += chunk.length;
+    this.offset += chunk.length;
+  }
+
+  write(chunk, encoding) {
+    const remainingSize = this.highWaterMark - this.offset;
+
+    if (chunk.length <= remainingSize) {
+      this[writeChunkSymbol](chunk, encoding);
+    } else {
+      const currentBufferChunk = chunk.slice(0, this.remainingSize);
+      this[writeChunkSymbol](currentBufferChunk);
+      this.queue.push(this.buffer);
+      this.buffer = Buffer.alloc(this.highWaterMark);
+      this.offset = 0;
+
+      const nextBufferChunk = chunk.slice(remainingSize);
+
+      if (nextBufferChunk.length < this.buffer.length) {
+        this[writeChunkSymbol](nextBufferChunk, encoding);
+      } else {
+        this.queue.push(nextBufferChunk);
+        this.writableLength += nextBufferChunk.length;
+      }
+    }
+  }
+
+  getBuffer() {
+    const bufferData = this.buffer.slice(0, this.offset);
+    this.queue.push(bufferData);
+    const ret = this.queue;
+    this.queue = [];
+    this.offset = 0;
+    this.writableLength = 0;
+    return ret;
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    defaultEncoding = 'utf-8',
+    highWaterMark = DEFAULT_HIGH_WATERMARK
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+
+    if (!this._write && !this._writev) {
+      throw new Error('[WRITE_AND_WRITEV_NOT_IMPLEMENTED]');
+    }
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.highWaterMark = highWaterMark;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(highWaterMark);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked) this.corked--;
+    if (this.corked === 0 && this.writableLength) {
+      this.writing = true;
+      this[doWriteSymbol](this.buffer.getBuffer());
+    }
+  }
+
+  setDefaultEncoding(encoding) {
+    this.defaultEncoding = encoding;
+  }
+
+  [errorFuncName](err, cb) {
+    if (typeof cb === 'function') {
+      process.nextTick(cb, err);
+    }
+    this.emit('error', err);
+  }
+
+  pipe() {
+    const err = new Error('[ERR_STREAM_CANNOT_PIPE]');
+    this[errorFuncName](err);
+  }
+
+  [validChunkSymbol](chunk, cb) {
+    let error;
+    if (chunk === null) {
+      error = new Error('[ERR_STREAM_NULL_VALUES]');
+    } else if (!Buffer.isBuffer(chunk)) {
+      error = new Error('[ERR_INVALID_ARG_TYPE]');
+    }
+
+    if (error) {
+      this[errorFuncName](error, cb);
+      return false;
+    }
+    return true;
+  }
+
+  write(chunk, encoding, cb) {
+    if (typeof encoding === 'function') {
+      cb = encoding;
+      encoding = null;
+    }
+
+    if (this.ended) {
+      const err = new Error('[ERR_STREAM_WRITE_AFTER_END]');
```

Comment:
Forgot about errors.

Why do you have errors as strings? Those should be potentially be separate errors classes and each have unique code.